galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,300 @@
+	pendingEvents int64
+
+	// lastSnapshotTime records the last time a snapshot was published.
+	lastSnapshotTime time.Time
+}
+
+// NewHandler creates a new Handler instance.
+func NewHandler(domainSuffix string, listener processing.Listener) *Handler {
+	pods, podsHandler := pod.NewCache()
+	nodes, nodesHandler := node.NewCache()
+
+	return &Handler{
+		domainSuffix:     domainSuffix,
+		listener:         listener,
+		services:         make(map[resource.FullName]resource.Entry),
+		endpoints:        make(map[resource.FullName]resource.Entry),
+		localityStrategy: locality.NewStrategy(pods, nodes),
+		podsHandler:      podsHandler,
+		nodesHandler:     nodesHandler,
+	}
+}
+
+// Handle incoming events and generate synthetic ServiceEntry projections.
+func (p *Handler) Handle(event resource.Event) {
+	if event.Kind == resource.FullSync {
+		p.distribute = true
+		p.notifyChanged()
+		return
+	}
+
+	switch event.Entry.ID.Collection {
+	case metadata.Endpoints.Collection:
+		// Update the projections
+		p.handleEndpointsEvent(event)
+	case metadata.Service.Collection:
+		// Update the projections
+		p.handleServiceEvent(event)
+	case metadata.Node.Collection:
+		// Just add the node to the cache.
+		p.nodesHandler.Handle(event)
+	case metadata.Pod.Collection:
+		// Just add the pod to the cache.
+		p.podsHandler.Handle(event)
+	default:
+		scope.Warnf("received event with unexpected collection: %v", event.Entry.ID.Collection)
+	}
+}
+
+// Builds the snapshot of the current resources.
+func (p *Handler) BuildSnapshot() snapshot.Snapshot {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	now := time.Now()
+	monitoring.RecordProcessorSnapshotPublished(p.pendingEvents, now.Sub(p.lastSnapshotTime))
+	p.lastSnapshotTime = now
+
+	b := snapshot.NewInMemoryBuilder()
+
+	entries := make([]*mcp.Resource, 0, len(p.mcpResources))
+	for _, r := range p.mcpResources {
+		entries = append(entries, r)
+	}
+	version := fmt.Sprintf("%d", p.version)
+	b.Set(collection.String(), version, entries)
+
+	p.pendingEvents = 0
+
+	return b.Build()
+}
+
+func (p *Handler) handleServiceEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the service's version has changed.
+		curEntry, ok := p.services[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("Received service for the current, known version: %v", event)
+			return
+		}
+
+		// Store the service.
+		p.services[fullName] = entry
+
+		// Get the associated endpoints, if available.
+		var endpoints *coreV1.Endpoints
+		if endpointsEntry, ok := p.endpoints[fullName]; ok {
+			endpoints = endpointsEntry.Item.(*coreV1.Endpoints)
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(entry, endpoints)
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
```

Comment:
The current implementation overwrites existing Endpoints in the ServiceEntry with the new list.

Reply:
yes, that's what we're doing here ... are you suggesting there's an issue with doing that?

Reply:
The old code has disappeared, I need to look at the new code, but yes, if my understanding is correct, upon receiving a new endpoint for a service with existing endpoints, it will override the existing endpoints with that single ep.

Reply:
@andraxylia I don't think your understanding is correct here. 

Here's my understanding... The k8s `Endpoints` contains the full list of endpoints for the service. When it is "added", then the service was just created and new endpoints are just now being spun up (the Endpoints contain the complete list).  If it's updated, the list has been modified and the event contains the new list. If it's deleted, the service is likely also being deleted shortly.

So if my understanding is correct, it's always correct to overwrite the old list of endpoints with the new.

Reply:
Also if it is empty it means the endpoints are restarting or gone, and you need to push the empty list. 

Yes, AFAIK your understanding is correct.

Reply:
Got it, we always get the full list of endpoints, the current informer will never send a single Endpoint.

cc @ayj for incremental MCP of Endpoints we may need to do the diff ourselves.